spring-integration-mqtt/src/test/java/org/springframework/integration/mqtt/ClientManagerBackToBackTests.java

```diff
@@ -0,0 +1,158 @@
+	}
+
+	private void testSubscribeAndPublish(Class<?> configClass) {
+		try (var ctx = new AnnotationConfigApplicationContext(configClass)) {
+			// given
+			var input = ctx.getBean("mqttOutFlow.input", MessageChannel.class);
+			var output = ctx.getBean("fromMqttChannel", PollableChannel.class);
+			String testPayload = "foo";
+
+			// when
+			input.send(MessageBuilder.withPayload(testPayload).setHeader(MqttHeaders.TOPIC, TOPIC_NAME).build());
+			Message<?> receive = output.receive(10_000);
+
+			// then
+			assertThat(receive).isNotNull();
+			assertThat(receive.getPayload()).isEqualTo(testPayload);
+		}
+	}
+
+	@Configuration
+	@EnableIntegration
+	public static class Mqttv3Config {
+
+		@Bean
+		public TaskScheduler taskScheduler() {
+			return new ThreadPoolTaskScheduler();
+		}
+
+		@Bean
+		public Mqttv3ClientManager mqttv3ClientManager(MqttPahoClientFactory pahoClientFactory) {
+			return new Mqttv3ClientManager(pahoClientFactory, SHARED_CLIENT_ID);
+		}
+
+		@Bean
+		public MqttPahoClientFactory pahoClientFactory() {
+			var pahoClientFactory = new DefaultMqttPahoClientFactory();
+			MqttConnectOptions connectionOptions = new MqttConnectOptions();
+			connectionOptions.setServerURIs(new String[]{ MosquittoContainerTest.mqttUrl() });
+			pahoClientFactory.setConnectionOptions(connectionOptions);
+			return pahoClientFactory;
+		}
+
+		@Bean
+		public IntegrationFlow mqttOutFlow(MqttPahoClientFactory pahoClientFactory,
+				Mqttv3ClientManager mqttv3ClientManager) {
+
+			var mqttHandler = new MqttPahoMessageHandler(SHARED_CLIENT_ID, pahoClientFactory);
+			mqttHandler.setClientManager(mqttv3ClientManager);
+			return f -> f.handle(mqttHandler);
+		}
+
+		@Bean
+		public IntegrationFlow mqttInFlow(MqttPahoClientFactory pahoClientFactory,
+				Mqttv3ClientManager mqttv3ClientManager) {
+
+			var mqttAdapter = new MqttPahoMessageDrivenChannelAdapter(SHARED_CLIENT_ID, pahoClientFactory, TOPIC_NAME);
+			mqttAdapter.setClientManager(mqttv3ClientManager);
+			return IntegrationFlow.from(mqttAdapter)
+					.channel(c -> c.queue("fromMqttChannel"))
+					.get();
+		}
+
+	}
+
+	@Configuration
+	@EnableIntegration
+	public static class Mqttv5Config {
+
+		@Bean
+		public TaskScheduler taskScheduler() {
+			return new ThreadPoolTaskScheduler();
+		}
+
+		@Bean
+		public Mqttv5ClientManager mqttv5ClientManager() {
+			return new Mqttv5ClientManager(MosquittoContainerTest.mqttUrl(), SHARED_CLIENT_ID);
+		}
+
+		@Bean
+		public IntegrationFlow mqttOutFlow(Mqttv5ClientManager mqttv5ClientManager) {
+			var mqttHandler = new Mqttv5PahoMessageHandler(MosquittoContainerTest.mqttUrl(), SHARED_CLIENT_ID);
+			mqttHandler.setClientManager(mqttv5ClientManager); // todo: add into ctor
+			return f -> f.handle(mqttHandler);
+		}
+
+		@Bean
+		public IntegrationFlow mqttInFlow(Mqttv5ClientManager mqttv5ClientManager) {
+			var mqttAdapter = new Mqttv5PahoMessageDrivenChannelAdapter(MosquittoContainerTest.mqttUrl(), SHARED_CLIENT_ID, TOPIC_NAME);
```

Comment:
I suppose, 
```java
var mqttAdapter = new Mqttv5PahoMessageDrivenChannelAdapter(MosquittoContainerTest.mqttUrl(), SHARED_CLIENT_ID, TOPIC_NAME);
mqttAdapter.setClientManager(mqttv5ClientManager);
```
is worse than
```java
var mqttAdapter = new Mqttv5PahoMessageDrivenChannelAdapter(clientManager);
```
@artembilan  - what do you think, does it make sense to remove clientId as required in abstract handler/adapters and create mutually exclusive ctors there as well?

Reply:
I thought that what we have agreed: or `clientId` for an internal client instantiation or `clientManager` which is going to control that client for us.